File: Python/367.ValidPerfectSquare.py
#Given a positive integer num, write a function which returns True if num is a perfect square else False.
class Solution(object):
    def isPerfectSquare(self, num):
        low=0
        high=num
        #Starting from zero till the number we need to check for perfect square
        while(low<=high):
            #Calulating middle value by using right shift operator
            mid=(low+high)>>1
            #If the square of the middle value is equal to the number then it is a perfect square else not
            if(mid*mid==num):
                return True
            #If the square of the middle value is less than the number we increment the low variable else the high variable is decremented.
            #The loop will continue till the low value becomes more than the high value or the number is a perfect square then True will be 
            #returned
            elif(mid*mid<num):
                low=mid+1
            else:
                high=mid-1
        return False        

File: C++/367. ValidPerfectSquare.cpp
/**
 * @file 367. ValidPerfectSquare.cpp
 * @author Vipul kumar Singh  ((https://github.com/vipul-2003))
 * @version 0.1
 * @date 2021-11-11
 * @link https://leetcode.com/problems/valid-perfect-square/
 * @copyright Copyright (c) 2021
 */

/*
Problems-
Given a positive integer num, write a function which returns True if num is a perfect square else False.
Follow up: Do not use any built-in library function such as sqrt.

*/

/*
Example 1:

Input: num = 16
Output: true

Example 2:

Input: num = 14
Output: false

Constraints:

1 <= num <= 2^31 - 1

*/

class Solution
{
public:
    bool isPerfectSquare(int num)
    {

        bool temp = false; // to check that the valid perfect square found or not , if yes , then now stop ... 

        if (num == 1) // edge cases
            return true;

        else if (num == 0) // edge cases
            return 0;

        else
        {
            for (long long int i = 2; i <= num / 2; i++) // for loop for traversing 
            {
                if (i * i == num) // i^2 == num (i square equal to the number "num")
                {
                    temp = true;
                    break;
                }
            }
        }

        return temp;
    }
};
